### PR TITLE
hierarchy improvements: Accelerated T-measures and L-measures

### PR DESCRIPTION
**Pull request by bmcfee**
This PR combines two modifications of the segment heirarchy module:
1. As mentioned in #201 , the implementation of T-measures can be accelerated significantly by using a more clever implementation of inversion counting.
2. The T-measures can be generalized by accepting segment labels (for each layer) rather than assigning each segment a unique identifier.  This allows us to compute hierarchical segment label agreement.  The more generic version is called `L-measure` (L for label and/or lattice, since the hierarchy is no longer a tree), and is derived in an upcoming article.

So far, this PR only includes regression tests to verify that the T-measures are still numerically equivalent after the re-implementation.  I still need to expand the tests to cover L-measure regressions, and unit tests for the helper functions.

### CODE REVIEW

**Comment by craffel** [comment]
Sweet, speedups are cool with regression tests.  For L-measure, I think we should wait until we have tests for it, or maybe even until it's published.  `mir_eval` convention is to only implement "standard" metrics for a task; we bent this rule a little for hierarchy eval, but I think having something which is not written down anywhere may bend it a bit too far.  What's the expected publication date?

**Comment by bmcfee** [comment]
> convention is to only implement "standard" metrics for a task; we bent this rule a little for hierarchy eval, but I think having something which is not written down anywhere may bend it a bit too far.

Agreed -- I'm in no rush to merge, but I wanted to get the PR open so we can track regressions.

Article should be out some time in 2017 ; submission deadline is at the end of this month.

**Comment by bmcfee** [comment]
Quick note: the regression tests for t-measures have slightly different values now, but this is entirely down to rounding error in the JSON output.  The new versions have one fewer digit, but are in fact identical when computed.

**Comment by craffel** [comment]
Overall looks reasonable to me. The T-measure accelerations are fine as long as they pass regression tests (json rounding aside).  The L-measure, I don't know what it is/what it computes, but I assume it's correct.  Is the pub out?  In October you wrote
> Article should be out some time in 2017; submission deadline is at the end of this month.

**Comment by bmcfee** [comment]
> Overall looks reasonable to me. The T-measure accelerations are fine as long as they pass regression tests (json rounding aside).

Yup, they do.  You can eyeball the diffs and see that the differences in values are only in the 15th digits.

>The L-measure, I don't know what it is/what it computes, but I assume it's correct. 

It's essentially the same as the T-measure applied to a different type of segment agreement matrix.  T-measure is actually just the special case where each segment within a level gets a unique label.

> Is the pub out? In October you wrote

Not yet. :cry: It's still under review.  I can send you the draft if you want to give it a skim.

**Comment by craffel** [comment]
Ok to wait til the pub comes out?  Maybe I'm wrong but it doesn't seem like people will be rushing to use this until they have had the chance to read about it.

**Comment by bmcfee** [comment]
> Ok to wait til the pub comes out? Maybe I'm wrong but it doesn't seem like people will be rushing to use this until they have had the chance to read about it.

Sure -- I just wanted to have the PR in place.  OTOH, if you're planning a release ahead of ismir, it might be nice to include since the t-measure accelerations could be useful.  It's totally your call though.

**Comment by craffel** [comment]
Should be able to do another release before ISMIR.  We can aim this to put this in there.

**Comment by bmcfee** [comment]
> another release before ISMIR

to be clear, I meant in advance of the deadline, so that people can use it for submissions.

**Comment by craffel** [comment]
I could see that happening.

**Comment by bmcfee** [comment]
Quick note on hobgoblins:

Could we at least move the pep8 check to after the test execution?  It's not helpful to have style failures precede correctness failures while developing.

**Comment by craffel** [comment]
> Could we at least move the pep8 check to after the test execution? It's not helpful to have style failures precede correctness failures while developing.

Sure, either way is fine.

**Comment by bmcfee** [comment]
The paper is now accepted, final print should appear soon: http://journal.frontiersin.org/article/10.3389/fpsyg.2017.01337/abstract

**Comment by craffel** [comment]
Congrats!  Where does this PR stand?  Last comment you were still working on tests for the L-measure version, but that was some time ago.  I assume that the coverage increasing indicates this is ready for review at least?

**Comment by bmcfee** [comment]
Yeah, it's done and ready for CR.

File: mir_eval/hierarchy.py [hunk 1]
```diff
@@ -245,6 +321,105 @@ def _gauc(ref_lca, est_lca, transitive, window):
...
+        if a[i] < b[j]:
+            i += 1
+        elif a[i] >= b[j]:
+            inversions += np.sum(a_counts[i:]) * b_counts[j]
+            j += 1
+
+    return inversions
+
+
+def _compare_frame_rankings(ref, est, transitive=False):
```
Comment (craffel) [inline mir_eval/hierarchy.py]: Requesting simple sanity-check unit test.

File: mir_eval/hierarchy.py [hunk 2]
```diff
@@ -245,6 +321,105 @@ def _gauc(ref_lca, est_lca, transitive, window):
     return score
 
 
+def _count_inversions(a, b):
```
Comment (craffel) [inline mir_eval/hierarchy.py]: Requesting simple sanity-check unit test.

File: mir_eval/hierarchy.py [hunk 3]
```diff
@@ -143,6 +167,65 @@ def _lca(intervals_hier, frame_size):
     return lca_matrix.tocsr()
 
 
+def _meet(intervals_hier, labels_hier, frame_size):
```
Comment (bmcfee) [inline mir_eval/hierarchy.py]: ok, I can do that.

Comment (bmcfee) [inline mir_eval/hierarchy.py]: done

File: mir_eval/hierarchy.py [hunk 4]
```diff
@@ -143,6 +167,65 @@ def _lca(intervals_hier, frame_size):
...
+    '''Compute the (sparse) least-common-ancestor (LCA) matrix for a
+    hierarchical segmentation.
+
+    For any pair of frames ``(s, t)``, the LCA is the deepest level in
+    the hierarchy such that ``(s, t)`` are contained within a single
+    segment at that level.
+
+    Parameters
+    ----------
+    intervals_hier : list of ndarray
```
Comment (bmcfee) [inline mir_eval/hierarchy.py]: done

File: mir_eval/hierarchy.py (hunk 2 above)
Comment (bmcfee) [inline mir_eval/hierarchy.py]: done

Comment (bmcfee) [inline mir_eval/hierarchy.py]: done

File: mir_eval/hierarchy.py (hunk 1 above)
Comment (bmcfee) [inline mir_eval/hierarchy.py]: done

File: mir_eval/hierarchy.py [hunk 5]
```diff
@@ -143,6 +169,69 @@ def _lca(intervals_hier, frame_size):
     return lca_matrix.tocsr()
 
 
+def _meet(intervals_hier, labels_hier, frame_size):
+    '''Compute the (sparse) least-common-ancestor (LCA) matrix for a
+    hierarchical segmentation.
+
+    For any pair of frames ``(s, t)``, the LCA is the deepest level in
+    the hierarchy such that ``(s, t)`` are contained within a single
```
Comment (craffel) [inline mir_eval/hierarchy.py]: Maybe add a note here (or elsewhere, wherever you think is appropriate) that levels start from level 1 (as opposed to level 0).

**Comment by craffel** [comment]
@urinieto can you look this over and confirm the changes match your expectations?  I don't have the expertise to sign off on the new metrics.

File: mir_eval/hierarchy.py (hunk 5 above)
Comment (bmcfee) [inline mir_eval/hierarchy.py]: I don't think that matters? All the comparisons are relative, so starting at 1 or 0 won't make a difference.

**Comment by bmcfee** [comment]
@urinieto ?

File: mir_eval/hierarchy.py [hunk 6]
```diff
@@ -210,29 +298,22 @@ def _gauc(ref_lca, est_lca, transitive, window):
...
 
-        # Compute normalization constant
-        normalizer = float(ref_rank.sum())
+        ref_score = np.concatenate((ref_score[:idx], ref_score[idx+1:]))
+        est_score = np.concatenate((est_score[:idx], est_score[idx+1:]))
 
-        # Add up agreement for frames
-        if normalizer > 0:
-            score += np.sum(np.logical_and(ref_rank, est_rank)) / normalizer
+        # If there are no comparisons to be made, move to the next frame
```
Comment (urinieto) [inline mir_eval/hierarchy.py]: This comment seems out of place. I don't think that's that what `_compare_frame_rankings` does. Also, I would remove the next empty line.

File: mir_eval/hierarchy.py [hunk 7]
```diff
@@ -98,6 +100,29 @@ def _hierarchy_bounds(intervals_hier):
...
+def _align_intervals(int_hier, lab_hier, t_min=0.0, t_max=None):
+    '''Align a hierarchical annotation to span a fixed start and end time.
+
+    Parameters
+    ----------
+    int_hier : list of list of intervals
+    lab_hier : list of list of str
+        Hierarchical segment annotations, encoded as a
+        list of list of intervals (int_hier) and list of
+        list of strings (lab_hier)
```
Comment (urinieto) [inline mir_eval/hierarchy.py]: The function arguments `t_min` and `t_max` are missing in the docstring.

File: mir_eval/hierarchy.py [hunk 8]
```diff
@@ -98,6 +100,29 @@ def _hierarchy_bounds(intervals_hier):
...
+
+    Returns
+    -------
+    intervals_hier, labels_hier
```
Comment (urinieto) [inline mir_eval/hierarchy.py]: I would clearly specify the types of the returned values. I would also enumerate them separately.

**Comment by bmcfee** [comment]
Ok, I think I've addressed Uri's comments, so this ought to be gtg.

**Comment by urinieto** [comment]
Sweet, thanks! Just approved my review.

**Comment by craffel** [comment]
Merged, thanks!

**Pull request merged**